File: Java/src/day14/day14.java
package day14;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;

public class day14 {
    public static void main(String[] args) throws IOException, NoSuchAlgorithmException {
        String player_instructions = get_input();
        System.out.println(part1(player_instructions, false));
        System.out.println(part1(player_instructions, true));
    }

    private static int part1(String instructions, boolean extended_hashing) throws NoSuchAlgorithmException {
        int counter = 0;
        List<Integer> valid_hashes = new ArrayList<>();
        Map<Integer, char[]> pre_calculated_hashes = new HashMap<>();
        while (true){
            char[] hashtext = calculate_hash(instructions, counter, extended_hashing, pre_calculated_hashes);
            Tuple2<Boolean, Character> triple_search_result = contains_triple(hashtext);
            if (triple_search_result.getFirst()) {
                for (int i = 0; i < 1000; i += 1){
                    char[] successive_hashtext = calculate_hash(instructions, counter + i + 1, extended_hashing, pre_calculated_hashes);
                    boolean five_search_result = contains_five(successive_hashtext, triple_search_result.getSecond());
                    if (five_search_result){
                        valid_hashes.add(counter);
                        break;
                    }
                }
            }
            counter += 1;
            if (valid_hashes.size() >= 64)
                break;
        }
        return valid_hashes.get(63);
    }

    private static char[] calculate_hash(String instructions, int counter, boolean extended_hashing, Map<Integer, char[]> pre_calculated_hashes) throws  NoSuchAlgorithmException {
        if (pre_calculated_hashes.containsKey(counter))
            return pre_calculated_hashes.get(counter);
        else {
            MessageDigest md = MessageDigest.getInstance("MD5");
            StringBuilder sb = new StringBuilder();
            sb.append(instructions);
            sb.append(counter);
            String to_be_hashed = sb.toString();
            String hashtext = generate_hash(md, to_be_hashed);
            if (extended_hashing) {
                for (int i = 0; i < 2016; i += 1)
                    hashtext = generate_hash(md, hashtext);
            }
            char[] hash_as_array = hashtext.toCharArray();
            pre_calculated_hashes.put(counter, hash_as_array);
            return hash_as_array;
        }
    }

    private static Tuple2<Boolean, Character> contains_triple(char[] hashtext){
        for (int i = 0; i < hashtext.length - 2; i += 1) {
            if (hashtext[i] == hashtext[i+1] && hashtext[i] == hashtext[i+2])
                return new Tuple2(true, hashtext[i]);
        }
        return new Tuple2(false, '0');
    }

    private static boolean contains_five(char[] hashtext, char seeked_match){
        for (int i = 0; i < hashtext.length - 4; i += 1){
            if (hashtext[i] == hashtext[i+1] && hashtext[i] == hashtext[i+2]
                    && hashtext[i] == hashtext[i+3] && hashtext[i] == hashtext[i+4] && hashtext[i] == seeked_match)
                return true;
        }
        return false;
    }

    private static String get_input() throws IOException {
        List<String> lines = Files.readAllLines(Paths.get("src/day14/input.txt"));
        return lines.get(0);
    }

    private static String generate_hash(MessageDigest md, String instructions){
        byte[] messageDigest = md.digest(instructions.getBytes());
        BigInteger no = new BigInteger(1, messageDigest);
        String hashtext = no.toString(16);
        while (hashtext.length() < 32) {
            hashtext = "0" + hashtext;
        }
        return hashtext;
    }

    public static class Tuple2<T,U>{
        private final T first;
        private final U second;

        public Tuple2(T first, U second) {
            this.first = first;
            this.second = second;
        }

        public T getFirst() { return first; }
        public U getSecond() { return second; }
    }
}

File: Python/day14/day14.py
import hashlib

def get_instructions():
    f = open("input.txt","r")
    for line in f:
        return line.strip()

def part1(instructions, extended_hashing):
    counter = 0
    valid_hash_pointers = []
    pre_calculated_hashes = {}
    while True:
        result = calculate_hash(instructions, counter, extended_hashing, pre_calculated_hashes)
        triple_search_result = contains_triple(result)
        if triple_search_result != -1:
            for i in range(1000):
                successive_hash = calculate_hash(instructions, counter + i + 1, extended_hashing, pre_calculated_hashes)
                if contains_five(successive_hash, triple_search_result):
                    valid_hash_pointers.append(counter)
                    break
        counter += 1
        if len(valid_hash_pointers) >= 64:
            return valid_hash_pointers[-1]

def calculate_hash(instructions, counter, extended_hashing, pre_calculated_hashes):
    if counter in pre_calculated_hashes:
        return pre_calculated_hashes[counter]
    else:
        str_counter = instructions + str(counter) 
        result = hashlib.md5(str_counter.encode("utf-8")).hexdigest()
        if extended_hashing == True:
            for i in range(2016):
                result = hashlib.md5(result.encode("utf-8")).hexdigest()
        pre_calculated_hashes[counter] = result
        return result


def contains_triple(result):
    for i in range(len(result) - 2):
        if (result[i] == result[i+1] == result[i+2]):
            return result[i]
    return -1

def contains_five(result, triple_search_result):
    for i in range(len(result) - 4):
        if (triple_search_result == result[i] == result[i+1] == result[i+2] == result[i+3] == result[i+4]):
            return True
    return False

if __name__ == "__main__":
    instructions = get_instructions()
    print(part1(instructions, False))
    print(part1(instructions, True))